### PR TITLE
docs: Evaluation module demo and documentation

### LINKED ISSUE #187: Notebook Demo and Docs for Eval Module
This ticket is for adding learning content and documentation about package Evaluation features, following up on the Evaluation module ticket and PR. 

See ticket #171 and PR #172

### PR DESCRIPTION
**Pull request by frayle-ons**
## ✨ Summary

These changes add documentation for the recent Evaluation module #171 #172 . Includes a new demo notebook `evaluation_workflow_demo.ipynb` walking users through the important concepts and code of the evaluation module. It also contains a new demo file of ground truth queries that work with the existing `fake_soc_dataset.csv` data, used to showcase evaluation.

It also has minor updates to the demo readme file.

## 📜 Changes Introduced

- [ ] (feat:) new demo notebook to showcase/explain the evaluation module
- [ ] (docs:) updated demo readme
- [ ] (feat:) new mock demo data to simulate ground truth queries for `fake_soc_dataset.csv` data

## ✅ Checklist

> **Please confirm you've completed these checks before requesting a review.**

- [ ] Code passes linting with **Ruff**
- [ ] Security checks pass using **Bandit**
- [ ] API and Unit tests are written and pass using **pytest**
- [ ] Terraform files (if applicable) follow best practices and have been validated (`terraform fmt` & `terraform validate`)
- [ ] DocStrings follow Google-style and are added as per Pylint recommendations
- [ ] Documentation has been updated if needed

## 🔍 How to Test

- Running the new notebook in a fresh install of the package to verify it works, and reading the content carefully to ensure it's correct. 
- Quickly browse the new mock data file
- and read through the updates to the readme documentation describing the new notebook

### CODE REVIEW

**Comment by jamie-ons** [comment]
# Overview
Have read through the notebook and think its really good and quite clear. Love the diagram.

I have:
- Run the new notebook in a fresh install of the package and can confirm it all runs without errors
- I have read the content and am mostly happy with the content.
- Read the new mock data file and am happy with it
- Read through the updates to the readme documentation describing the new notebook

# Proposed changes
There are a few changes that I think would improve the demo:

**Spelling**
This would help it read better in some places, here is a few examples, there are more in the notebook.
Typo | Correction
 -- | --
provies | provides
mulit-class | multi-class
peformance | performance
mutiple | multiple
there names | their names
Typing this all together | Tying this all together
fake_soc_daraset.csv | fake_soc_dataset.csv

If you could go through and check the spellings that would be great :)

## VectorStore creation
I think setting `overwrite=True` to all VectorStore creations would be better as it means if you rerun the cell for whatever reason, you don't end up getting the error:
```
Output directory already exists. Pass overwrite=True to overwrite the folder. 
```
This exists for the
- demo_vectorstore_full
- demo_vectorstore_half

## partially missing comment
```
results = evaluator.evaluate(
    vectorstores=[demo_vectorstore_full, demo_vectorstore_half],
    vectorstore_names=["full data vectorstore", "half data vectorstore"],
    output_file="./classifai_temp/demo_eval_results.csv",  # leaving this line blank will save the results to
)
```
I believe this comment should say 
```
    output_file="./classifai_temp/demo_eval_results.csv",  # leaving this line blank will save the results to evaluation_results.csv
```

## DEMO/README.md
This all looks good apart from the Note at the bottom is massive.
image

I would change it to the following to avoid this
```
> **Note:** The Evaluation module is currently in development and its API is subject to change in future releases.
```

File: DEMO/README.md [hunk 1]
```diff
@@ -66,6 +66,25 @@ It covers:
...
+
+* Creating multiple VectorStore instances with varying data coverage to showcase performance differences.
+
+* Instantiating an `Evaluation` object with ground truth data and selected metrics.
+
+* Running the `evaluate()` method to compute metrics across multiple VectorStores.
+
+* Memory-efficient evaluation using callable functions to load VectorStores on-demand, useful when evaluating many or large VectorStores.
+
+**Note:** The Evaluation module is currently in development and its API is subject to change in future releases.
```
Comment (jamie-ons) [inline DEMO/README.md]:
```suggestion
> **Note:** The Evaluation module is currently in development and its API is subject to change in future releases.
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    "# ClassifAI Evaluation Module - Overview and Usage\n",
+    "\n",
+    "---"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The Evaluation module provies a toolkit to evaluate the performance of VectorStores in a mulit-class, single-label classification setting. Provided the user has:\n",
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "The Evaluation module provides a toolkit to evaluate the performance of VectorStores in a multi-class, single-label classification setting. Provided the user has:\n",
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    "\n",
+    "- a constructed VectorStore (or multiple VectorStores) built from historically labeled (or similar) data;\n",
+    "- a held out collection of labelled ground truth data, not in the VectorStore;\n",
+    "\n",
+    "then this module can be used to evaluate VectorStore peformance, presenting results with a variety of available metrics that can be specified by the user."
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "then this module can be used to evaluate VectorStore performance, presenting results with a variety of available metrics that can be specified by the user."
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Multi-Class Single-Label Evaluation\n",
+    "\n",
+    "Currently the evaluation module only evaluates single label predictions meaning that, while ClassifAI is designed to return a ranked list of several semantically similar candidate entries to a provided query sample, only the top result will be considered when comparing the VectorStore result to a ground truth label provided by a user.\n",
+    "\n",
+    "![top_1_eval_image](files/eval_top_1_diagram.png)\n",
+    "\n",
+    "The Evaluation module is currently in development, and in the future its feature set may be extended to include a broader range of evaluation tasks such as multi-class multi-label classification, where potentially multiple labels for a ground truth sample can be compared and evaluated against mutiple ranked candidate predictions of the VectorStore."
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "The Evaluation module is currently in development, and in the future its feature set may be extended to include a broader range of evaluation tasks such as multi-class multi-label classification, where potentially multiple labels for a ground truth sample can be compared and evaluated against multiple ranked candidate predictions of the VectorStore."
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For multi-class, single-label evaluation we have implemented several metrics that can be calculated, there names and descriptions are as follows: \n",
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "For multi-class, single-label evaluation we have implemented several metrics that can be calculated, their names and descriptions are as follows: \n",
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+    "\n",
+    "- An introduction to the Evaluation Module and its main class `Evaluation`\n",
+    "\n",
+    "- Use of the demo file `fake_soc_eval_queries.csv` which is in the `DEMO/data/` repo folder. This data contains ground truth samples related to the demo knowledgebase CSV file `DEMO/data/fake_soc_dataset.csv`\n",
+    "\n",
+    "- Creating several VectorStores using the 'fake_soc_dataset.csv' file, using different amounts of data from the file to create several VectorStores of varying quality.\n",
+    "\n",
+    "- Setting up an Evaluation task using the fake queries file and a list of metrics to evaluate.\n",
+    "\n",
+    "- Typing this all together, evaluating the VectorStores against the ground truth query file using the evaluation module.\n",
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "- Tying this all together, evaluating the VectorStores against the ground truth query file using the evaluation module.\n",
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+   "metadata": {},
+   "source": [
+    "### Building VectorStores to Evaluate\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To begin, we're going to create 2 VectorStores from our `fake_soc_daraset.csv` file which contains <b>mock</b> SOC survey responses and their corresponding occupation codes. One VectorStore will be built from the full dataset, and the second one will be built from half the dataset. \n",
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "To begin, we're going to create 2 VectorStores from our `fake_soc_dataset.csv` file which contains <b>mock</b> SOC survey responses and their corresponding occupation codes. One VectorStore will be built from the full dataset, and the second one will be built from half the dataset. \n",
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+    "    vectoriser=demo_vectoriser,\n",
+    "    output_dir=\"./classifai_temp/half_vectorStore/\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If the code cells in this section ran succesfully then we now have 2 VectorStores we can use to evaluate with the Evaluation module."
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "If the code cells in this section ran successfully then we now have 2 VectorStores we can use to evaluate with the Evaluation module."
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "ClassifAI provides an `Evaluation` class (in the Evaluation module) which can be used to load a collection of ground truth queries and specify metrics to use. The evaluator constructor accepts 4 arguments:\n",
+    "\n",
+    "- A pandas dataframe with `['text', 'label']` columns/headers both of type string, which represent the text sample queries and gold standard ground truth label respectively,\n",
+    "- A list of evaluation metric names which must be strings correpsonding to one of the current avaialble metrics: `['accuracy', 'macro_recall', 'macro_precision', 'macro_f1']`,\n",
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "- A list of evaluation metric names which must be strings corresponding to one of the current available metrics: `['accuracy', 'macro_recall', 'macro_precision', 'macro_f1']`,\n",
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+   "metadata": {},
+   "source": [
+    "### Efficient VectorStore Loading and other settings"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We've already demonstrated the core featrures of the Evaluation module. This final section will show some additional implemeted features including:\n",
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "We've already demonstrated the core features of the Evaluation module. This final section will show some additional implemented features including:\n",
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+    "second_results"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Thats It!\n",
+    "\n",
+    "The results (this time with more metrics) from the last evaluation run should be saved to the file in the current directory as `evaluation_results.csv`, since we didn't specify a file name this time to save the resutls to. Remember you can disable writing results to file by setting `save_output=False` in the Evaluation constructor.\n",
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "The results (this time with more metrics) from the last evaluation run should be saved to the file in the current directory as `evaluation_results.csv`, since we didn't specify a file name this time to save the results to. Remember you can disable writing results to file by setting `save_output=False` in the Evaluation constructor.\n",
```

File: DEMO/evaluation_workflow_demo.ipynb [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "results = evaluator.evaluate(\n",
+    "    vectorstores=[demo_vectorstore_full, demo_vectorstore_half],\n",
+    "    vectorstore_names=[\"full data vectorstore\", \"half data vectorstore\"],\n",
+    "    output_file=\"./classifai_temp/demo_eval_results.csv\",  # leaving this line blank will save the results to\n",
```
Comment (jamie-ons) [inline DEMO/evaluation_workflow_demo.ipynb]:
```suggestion
    "    output_file=\"./classifai_temp/demo_eval_results.csv\",  # leaving this line blank will save the results to evaluation_results.csv\n",
```

**Comment by frayle-ons** [comment]
I've applied additional changes based on feedback. I've set all the VectorStores instantiations in the eval notebook to use `skip_save=True` so that no data is saved to file. This is fine since we don't actually load anything back in and resolves some of Jamie's review comments per using `overwrite=True`

**Review by jamie-ons — DISMISSED** [review]
Yes, had a chat to luke and skip_save is a better idea.

Have had another look at this and it all looks good now :)

File: DEMO/evaluation_workflow_demo.ipynb [hunk 13]
```diff
@@ -0,0 +1,381 @@
...
+    "### Building VectorStores to Evaluate\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To begin, we're going to create 2 VectorStores from our `fake_soc_dataset.csv` file which contains <b>mock</b> SOC survey responses and their corresponding occupation codes. One VectorStore will be built from the full dataset, and the second one will be built from half the dataset. \n",
+    "\n",
+    "Since the second VectorStore will contain only have the training data, we can reason that this lack of coverage will showcase poorer performance against a evaluation dataset that assesses the full coverage of the training data."
```
Comment (lukeroantreeONS) [inline DEMO/evaluation_workflow_demo.ipynb]: "only have half of the training data"

Comment (lukeroantreeONS) [reply DEMO/evaluation_workflow_demo.ipynb]: "will not perform as well as the full dataset due to lack of coverage"

File: DEMO/evaluation_workflow_demo.ipynb [hunk 14]
```diff
@@ -0,0 +1,381 @@
...
+    "    vectoriser=demo_vectoriser,\n",
+    "    skip_save=True,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If the code cells in this section ran successfully then we now have 2 VectorStores we can use to evaluate with the Evaluation module."
```
Comment (lukeroantreeONS) [inline DEMO/evaluation_workflow_demo.ipynb]: Suggest cutting this line - it makes it seem like we're not confident it works

File: DEMO/evaluation_workflow_demo.ipynb [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+    "    vectorstore_names=[\"full data vectorstore\", \"half data vectorstore\"],\n",
+    "    output_file=\"./classifai_temp/demo_eval_results.csv\",  # leaving this line blank will save the results to evaluation_results.csv\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The results object is a dataframe with provided VectorStore names as the row indexes, and each column is associated with a give metric. We should also see the results have been saved to a CSV file in the specified directory."
```
Comment (lukeroantreeONS) [inline DEMO/evaluation_workflow_demo.ipynb]: given metric

Comment (lukeroantreeONS) [reply DEMO/evaluation_workflow_demo.ipynb]: "saved to the specified output CSV file"

**Review by lukeroantreeONS — CHANGES_REQUESTED** [review]
This looks good - noted a few small tweaks to be made before merging

**Review by lukeroantreeONS — DISMISSED** [review]
Happy to approve 👍

**Review by rileyok-ons — APPROVED** [review]
Docstrings look great
